xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,79 +240,243 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
+        prefixRanges.add(new CidrRange(range.getAddressPrefix(), range.getPrefixLen().getValue()));
+      }
+      for (io.envoyproxy.envoy.config.core.v3.CidrRange range
+          : proto.getSourcePrefixRangesList()) {
+        sourcePrefixRanges.add(
+            new CidrRange(range.getAddressPrefix(), range.getPrefixLen().getValue()));
+      }
+    } catch (UnknownHostException e) {
+      throw new ResourceInvalidException("Failed to create CidrRange", e);
+    }
+    ConnectionSourceType sourceType;
+    switch (proto.getSourceType()) {
+      case ANY:
+        sourceType = ConnectionSourceType.ANY;
+        break;
+      case EXTERNAL:
+        sourceType = ConnectionSourceType.EXTERNAL;
+        break;
+      case SAME_IP_OR_LOOPBACK:
+        sourceType = ConnectionSourceType.SAME_IP_OR_LOOPBACK;
+        break;
+      default:
+        throw new ResourceInvalidException("Unknown source-type: " + proto.getSourceType());
+    }
+    return new FilterChainMatch(
+        proto.getDestinationPort().getValue(),
+        prefixRanges,
+        proto.getApplicationProtocolsList(),
+        sourcePrefixRanges,
+        sourceType,
+        proto.getSourcePortsList(),
+        proto.getServerNamesList(),
+        proto.getTransportProtocol());
+  }
 
+  @VisibleForTesting
+  static io.grpc.xds.HttpConnectionManager parseHttpConnectionManager(
+      HttpConnectionManager proto, boolean parseHttpFilter)
+      throws ResourceInvalidException {
     // Obtain max_stream_duration from Http Protocol Options.
     long maxStreamDuration = 0;
-    if (hcm.hasCommonHttpProtocolOptions()) {
-      HttpProtocolOptions options = hcm.getCommonHttpProtocolOptions();
+    if (proto.hasCommonHttpProtocolOptions()) {
+      HttpProtocolOptions options = proto.getCommonHttpProtocolOptions();
       if (options.hasMaxStreamDuration()) {
         maxStreamDuration = Durations.toNanos(options.getMaxStreamDuration());
       }
     }
 
-    // Parse filters.
-    List<NamedFilterConfig> filterChain = null;
-    if (parseFilter) {
-      filterChain = new ArrayList<>();
-      List<io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter>
-          httpFilters = hcm.getHttpFiltersList();
+    // Parse http filters.
+    List<NamedFilterConfig> filterConfigs = null;
+    if (parseHttpFilter) {
+      filterConfigs = new ArrayList<>();
+      Set<String> names = new HashSet<>();
       for (io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
-               httpFilter : httpFilters) {
+               httpFilter : proto.getHttpFiltersList()) {
         String filterName = httpFilter.getName();
+        if (!names.add(filterName)) {
+          throw new ResourceInvalidException(
+              "HttpConnectionManager contains duplicate HttpFilter: " + filterName);
+        }
         StructOrError<FilterConfig> filterConfig = parseHttpFilter(httpFilter);
```

Comment:
Looking at the code of `parseHttpFilter` (and then `parseRawFilterConfig`) I notice that this new code will now allow the FaultFilter (`type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault`) on the server side because of the use of the common FilterRegistry. Even though we don't support FaultFilter on the server side. This is a regression in the server side validation code

Reply:
Currently, the only supported common part is _reading in the HttpFilter configs_. That is, if a `type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault` is found in server side HttpConnectionManager, it will be read in and parsed. Parsing HttpFilter configs should be client/server-agnostic, as a filter or filter config is completely defined by the typeUrl of the proto, not how it is being used. Being able to parse the filter config doesn't necessarily apply the filter. Take the fault filter as an example. If an HttpConnectionManager contains a `type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault` configuration, we look up the common filter registry and find that it is supported by gRPC. So this filter config will be parsed. The next step is to turn this parsed filter config to a ClientInterceptor/ServerInterceptor that can be used to intercept client/server calls. On the client side, this is done by [looking up the common registry to find gRPC's implementation for FaultFilter](https://github.com/grpc/grpc-java/blob/fa4b980e07400a66cfaf3693785f873031b20482/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L411). Then we [check if this implementation supports client side's usage](https://github.com/grpc/grpc-java/blob/fa4b980e07400a66cfaf3693785f873031b20482/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L413). If it does, we turn the filter into a ClientInterceptor to be used by the Channel. Similarly for server side, after receiving _FaultFilter config_, the server we will look up the common registry to get the _filter implementation_, then if the implementation supports server side's usage (aka, if it implements ServerInterceptorBuilder), we will turn it to a ServerInterceptor to be used by the server.

This is a very pluggable design that decouples the input of filter config and output of ClientInterceptor/ServerInterceptor based on gRPC's own decision. It maximizes the amount of code/logic reuse while preserves the isolation of whether to support some type of HttpFilter on client/server side independently.

/cc @dapengzhang0 if you have any feeling/comment for this design based on your experience of implementing fault injection.

Reply:
> if you have any feeling/comment for this design based on your experience of implementing fault injection.

Yes, that's designed that way. The design also mentioned this: 
"gRPC will not fail validation if a map entry uses a config for a filter that is supported on only the gRPC client but is used on the gRPC server, or vice-versa. This is because the xDS client code cannot know whether a given RDS update will be used by a gRPC client or server."

Reply:
> This is because the xDS client code cannot know whether a given RDS update will be used by a gRPC client or server.

That's not true. We have `parseServerSideListener` because we know it's for the server side :-)

Reply:
> Being able to parse the filter config doesn't necessarily apply the filter.
...
> Similarly for server side, after receiving FaultFilter config, the server we will look up the common registry to get the filter implementation, then if the implementation supports server side's usage (aka, if it implements ServerInterceptorBuilder), we will turn it to a ServerInterceptor to be used by the server.

So the control plane sends you the faultFilter included in the listener on the server side and is to be used (based on matches) for the current connection. Because there is no server side implementation you are going to silently drop it and not do fault-injection even if the control plane asked you to do so? If it's okay here would it be okay elsewhere?

Reply:
> If an HttpConnectionManager contains a type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault configuration, we look up the common filter registry ... ... This is a very pluggable design that decouples the input of filter config and output of ClientInterceptor/ServerInterceptor based on gRPC's own decision. It maximizes the amount of code/logic reuse while preserves the isolation of whether to support some type of HttpFilter on client/server side independently.

So the implementation defines the correct behavior instead of getting the implementation to match the correct/expected behavior?

Reply:
>Because there is no server side implementation you are going to silently drop it and not do fault-injection even if the control plane asked you to do so? 

Oh, this comment is on handling LDS response, not RDS response. Then the [following bullet point](https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md#top-level-filter-configs) in the spec answers the question:

- Every filter config in the list must be of a type (as identified by the type_url field) that is present in the filter registry. Note that the set of known filter config types will be different on the gRPC client and server; if the HttpConnectionManager proto is inside an HTTP API Listener, it will look only at filters registered for the gRPC client, whereas if it is inside a TCP Listener, it will look only at filters registered for the gRPC server. Any unknown config type will cause the Listener resource to be NACKed, unless the is_optional field is true, in which case the individual filter will be ignored.

Reply:
> It is implementing the expected behavior, but you lack the knowledge of how gRPC supports HttpFilter. See https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md

So I looked at https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md#top-level-filter-configs (which BTW can also have bugs), and it says

"When gRPC receives ... it will validate the list of filters as follows:
...
 if the HttpConnectionManager proto is inside an HTTP API Listener, it will look only at filters registered for the gRPC client, whereas if it is inside a TCP Listener, it will look only at filters registered for the gRPC server. 
..."

So for this to be implemented you will have to change the code in `parseRawFilterConfig` at line 545 to check if this filter is client/server type after passing down client-side or server-side from the top.

Reply:
Yes, that's what I am changing now. It will need to carry around a `isForClient` bit, which is pretty annoying...

Reply:
> Yes, that's what I am changing now. It will need to carry around a `isForClient` bit, which is pretty annoying...

Oh, okay. In that case I will resume the review once you commit your next set of changes

Reply:
Updated the implementation. Parsing the top-level HttpFilter config (i.e., HttpFilters in HttpConnectionManager) requires knowing if the specific HttpFilter is supported for client-side or server side, but we don't/cannot distinguish the usage for parsing override HttpFilter configs, as we are not able to know if the enclosing RouteConfiguration is for a client API listener or for a server TCP listener.